feat: sshnp_docker_demo

JeremyTubongbanua opened the pull request:
## What I did

- demo of sshnp between two docker containers (+1 more for sshrvd)

closes #96

JeremyTubongbanua committed:
docs: updated at_cookbook README.md

JeremyTubongbanua committed:
chore: add `*.atKeys` to `.gitignore`

JeremyTubongbanua committed:
create: dirs `sshnp`, `sshnpd`, and `sshrvd`

JeremyTubongbanua committed:
docs: `sshnp_docker_demo` readme

JeremyTubongbanua committed:
feat: `sshnp`, `sshnpd`, `sshrvd` Dockerfiles

JeremyTubongbanua committed:
feat: shell scripts

JeremyTubongbanua committed:
chore: delete `tmp` files

JeremyTubongbanua committed:
feat: keys placeholder file & copies only keys over to container

JeremyTubongbanua committed:
docs: README

natt-n committed:
docs: step1docs fix

JeremyTubongbanua committed:
fix: container name

JeremyTubongbanua committed:
feat: shell script place holders

JeremyTubongbanua committed:
docs: sshnp_docker_demo readme finished i think

JeremyTubongbanua committed:
docs: added docker improvement

JeremyTubongbanua committed:
Revert "docs: updated at_cookbook README.md"

This reverts commit d844fcc022b95cdf587b21494a8d5057a49ad2a6.

JeremyTubongbanua committed:
revert: `.gitignore` random edit

JeremyTubongbanua commented:
@natt-n followed the `README.md` and got it working!

JeremyTubongbanua committed:
chore: `<ip>` placeholder

JeremyTubongbanua committed:
docs: code block format

JeremyTubongbanua committed:
docs: fix prerequisites

JeremyTubongbanua committed:
docs: edit start

cpswan reviewed (changes requested):
Rather than having fiddly instructions for people to specify their architecture the Dockerfiles should determine the architecture and act accordingly. When doing `RUN` commands this can generally be achieved with a `case` statement based on `uname -m` e.g. we use this for the [Dockerfile.package in sshnoports](https://github.com/atsign-foundation/sshnoports/blob/trunk/Dockerfile.package):

```dockerfile
RUN set -eux; \
    case "$(dpkg --print-architecture)" in \
        amd64) \
            ARCH="x64";; \
        armhf) \
            ARCH="arm";; \
        arm64) \
            ARCH="arm64";; \
        riscv64) \
            ARCH="riscv64";; \
    esac; \
    mkdir sshnp; \
    mkdir tarball; \
    dart pub get; \
    dart compile exe bin/activate_cli.dart -v -o sshnp/at_activate; \
    dart compile exe bin/sshnp.dart -v -o sshnp/sshnp; \
    dart compile exe bin/sshnpd.dart -v -o sshnp/sshnpd; \
    dart compile exe bin/sshrv.dart -v -o sshnp/sshrv; \
    dart compile exe bin/sshrvd.dart -v -o sshnp/sshrvd; \
    cp scripts/* sshnp; \
    tar -cvzf tarball/sshnp-linux-${ARCH}.tgz sshnp
```

It's also possible to use the `TARGETPLATFORM` ARG variable that gets passed in to a build [like this](https://github.com/BretFisher/multi-platform-docker-build#add-dockerfile-logic-to-detect-the-platform-it-needs-to-use):

```dockerfile
ARG TARGETPLATFORM
RUN case ${TARGETPLATFORM} in \
         "linux/amd64")  ARCH=amd64  ;; \
         "linux/arm64")  ARCH=arm64  ;; \
         "linux/arm/v7") ARCH=armhf  ;; \
         "linux/arm/v6") ARCH=armel  ;; \
         "linux/386")    ARCH=i386   ;; \
    esac \
 && wget -q https://github.com/org/thing/releases/download/${VERSION}/whatever-${ARCH} -O /thing
```

More generally it's considered good practice to have multi line `RUN` statements rather than many of them, as each command in a Dockerfile creates a new layer in the resulting image.

JeremyTubongbanua committed:
feat: docker build template shell script

JeremyTubongbanua committed:
docs: Dockerfile section of README

JeremyTubongbanua committed:
docs: update what is in each dir

JeremyTubongbanua committed:
docs: remove Dockerfile edit step in Prereq

JeremyTubongbanua committed:
feat: one dockerfile instead of three

JeremyTubongbanua commented:
> Rather than having fiddly instructions for people to specify their architecture the Dockerfiles should determine the architecture and act accordingly. When doing `RUN` commands this can generally be achieved with a `case` statement based on `uname -m` e.g. we use this for the [Dockerfile.package in sshnoports](https://github.com/atsign-foundation/sshnoports/blob/trunk/Dockerfile.package):
> 
> ```dockerfile
> RUN set -eux; \
>     case "$(dpkg --print-architecture)" in \
>         amd64) \
>             ARCH="x64";; \
>         armhf) \
>             ARCH="arm";; \
>         arm64) \
>             ARCH="arm64";; \
>         riscv64) \
>             ARCH="riscv64";; \
>     esac; \
>     mkdir sshnp; \
>     mkdir tarball; \
>     dart pub get; \
>     dart compile exe bin/activate_cli.dart -v -o sshnp/at_activate; \
>     dart compile exe bin/sshnp.dart -v -o sshnp/sshnp; \
>     dart compile exe bin/sshnpd.dart -v -o sshnp/sshnpd; \
>     dart compile exe bin/sshrv.dart -v -o sshnp/sshrv; \
>     dart compile exe bin/sshrvd.dart -v -o sshnp/sshrvd; \
>     cp scripts/* sshnp; \
>     tar -cvzf tarball/sshnp-linux-${ARCH}.tgz sshnp
> ```
> 
> It's also possible to use the `TARGETPLATFORM` ARG variable that gets passed in to a build [like this](https://github.com/BretFisher/multi-platform-docker-build#add-dockerfile-logic-to-detect-the-platform-it-needs-to-use):
> 
> ```dockerfile
> ARG TARGETPLATFORM
> RUN case ${TARGETPLATFORM} in \
>          "linux/amd64")  ARCH=amd64  ;; \
>          "linux/arm64")  ARCH=arm64  ;; \
>          "linux/arm/v7") ARCH=armhf  ;; \
>          "linux/arm/v6") ARCH=armel  ;; \
>          "linux/386")    ARCH=i386   ;; \
>     esac \
>  && wget -q https://github.com/org/thing/releases/download/${VERSION}/whatever-${ARCH} -O /thing
> ```
> 
> More generally it's considered good practice to have multi line `RUN` statements rather than many of them, as each command in a Dockerfile creates a new layer in the resulting image.

Thank you, @cpswan, 

finished in [74ef5e1](https://github.com/atsign-foundation/at_demos/pull/97/commits/74ef5e18be460c6ed5c9688927a9f16a1908e8cf)

JeremyTubongbanua commented:
Also made a few new changes

- there is one Dockerfile in the root of the project now, (instead of 3 in each directory). The `db.sh` script will copy over the Dockerfile and remove it once done with it. Now we only need to edit one Dockerfile
- there is a `db-template.sh` script now, each of the there `db.sh`'s will call this script with an argument
- README improvements

cpswan commented on sshnp_docker_demo/Dockerfile:
```diff
@@ -0,0 +1,42 @@
+    riscv64) \
+    ARCH="riscv64";; \
+    esac; \
+    wget https://github.com/atsign-foundation/sshnoports/releases/download/v3.1.2/sshnp-linux-${ARCH}.tgz ; \
```
version shouldn't be hard coded, but rather passed as an ARG or determined by doing a request to find the latest version e.g.

```
VERSION=$(curl --silent "https://api.github.com/repos/atsign-foundation/sshnoports/releases/latest"
 | grep -Po '"tag_name": "v\K.*?(?=")')
 ```

JeremyTubongbanua replied:
done in https://github.com/atsign-foundation/at_demos/pull/97/commits/d9e95e7c2709c1ccc9d41eda807259d3e2e5dcea

cpswan commented on sshnp_docker_demo/Dockerfile:
```diff
@@ -0,0 +1,42 @@
+
+RUN touch ${HOMEDIR}/.ssh/authorized_keys
+
+COPY keys/*.atKeys ${HOMEDIR}/.atsign/keys
```
Do the COPYs first, and then the RUNs can all be in one block.

JeremyTubongbanua replied:
Done in https://github.com/atsign-foundation/at_demos/pull/97/commits/14c5d7295aea135f8e9a9d6a1622c4e78bee4b40

cpswan commented on sshnp_docker_demo/Dockerfile:
```diff
@@ -0,0 +1,42 @@
+FROM ubuntu
```
Should be pinned to a particular SHA tag.

We tend to prefer debian:slim as the base for this type of image.

JeremyTubongbanua replied:
Need help with this- 

This is the image I am using:
```
FROM debian:stable-20230522-slim@sha256:d828cca5497a2519da9c6d42372066895fa28a69f1e8a46a38ce8f750bd2adf0
```

I get an error here:
```
#0 0.179 sysctl: permission denied on key "net.ipv4.ping_group_range"
------
Dockerfile:35
--------------------
  33 |     RUN touch ${HOMEDIR}/.ssh/authorized_keys
  34 |     RUN addgroup --gid ${GROUP_ID} ${USER} 
  35 | >>> RUN sysctl -w net.ipv4.ping_group_range="0 1024" 
  36 |     RUN useradd --system --uid ${USER_ID} --gid ${GROUP_ID} --shell /bin/bash --home ${HOMEDIR} ${USER} 
  37 |     RUN usermod -aG sudo ${USER} 
--------------------
ERROR: failed to solve: process "/bin/sh -c sysctl -w net.ipv4.ping_group_range=\"0 1024\"" did not complete successfully: exit code: 255
```

I don't run into this issue when I go back to `FROM ubuntu` though

cpswan replied:
What are we doing `sysctl -w net.ipv4.ping_group_range="0 1024"` for?

JeremyTubongbanua replied:
Am not sure, I took it from [here](https://github.com/atsign-foundation/sshnoports/blob/fa8908fa3088d38606f2c883e950870c0e0087fe/Dockerfile#L20)

When I comment out the line, I get this error on `sshnp`:

```
Remote sshnpd error: Remote SSH Client failure : SocketException: 
Connection refused (OS Error: Connection refused, errno = 111), address = 172.17.0.2, port = 51710
```

Works fine when I uncomment.

(This is all using `FROM ubuntu` as well)

JeremyTubongbanua replied:
I still also run into the issue up above where the docker image does not build because of permission errors when using `FROM debian:stable-20230522-slim@sha256:d828cca5497a2519da9c6d42372066895fa28a69f1e8a46a38ce8f750bd2adf0`

cpswan commented on sshnp_docker_demo/db-template.sh:
```diff
@@ -0,0 +1,21 @@
+#!/bin/bash
+
+#
+# DO NOT RUN THIS SCRIPT. THIS IS MEANT TO BE RAN IN ONE OF THE THERE DIRECTORIES
```
meant to be RUN?
THERE?

JeremyTubongbanua replied:
fixed in https://github.com/atsign-foundation/at_demos/pull/97/commits/8b417a9cf87bcdfc6242f164b030e6f75e7195eb

cpswan replied:
RAN is still not right, it should be RUN

JeremyTubongbanua replied:
https://github.com/atsign-foundation/at_demos/pull/97/commits/455e2c955fcdb31fa6876a381be79209897ad466

cpswan reviewed (changes requested):
A few more ways to improve the Dockerfile (see inline comments).

JeremyTubongbanua committed:
chore: `db-template.sh` comment

JeremyTubongbanua committed:
feat: use dynamic version

JeremyTubongbanua committed:
chore: comment chmods

JeremyTubongbanua committed:
chore: move copies to top

JeremyTubongbanua committed:
feat: one `run` dockerfile

cpswan commented on sshnp_docker_demo/Dockerfile:
```diff
@@ -0,0 +1,41 @@
+
+RUN apt-get update && apt-get install -y openssh-server wget sudo iputils-ping iproute2 ncat telnet net-tools nmap iperf3 tmux traceroute nano vim watch curl dpkg ; \
+    touch ${HOMEDIR}/.ssh/authorized_keys ; \
+    set -eux; \
```
`set -eux` should be the first line of the `RUN`

JeremyTubongbanua replied:
Like this? https://github.com/atsign-foundation/at_demos/pull/97/commits/ce76b968f65b4163ce77bd5c1143671934b192ba

JeremyTubongbanua committed:
chore: set -eux first line of run

JeremyTubongbanua committed:
chore: RAN -> RUN

cpswan committed:
chore: Tidy Dockerfile

cpswan committed:
fix: Reorder file ops

cpswan committed:
fix: Comment out sysctl line

cpswan commented:
@JeremyTubongbanua take a look at https://github.com/atsign-foundation/at_demos/blob/cpswan-dockerfile-tidy/sshnp_docker_demo/Dockerfile for what I think the Dockerfile **should** look like.

You can always PR that branch into this one (or just merge it, or copy/paste).

JeremyTubongbanua commented:
> @JeremyTubongbanua take a look at https://github.com/atsign-foundation/at_demos/blob/cpswan-dockerfile-tidy/sshnp_docker_demo/Dockerfile for what I think the Dockerfile **should** look like.
> 
> You can always PR that branch into this one (or just merge it, or copy/paste).

I get this error when running `sshnp`

```
PathAccessException: Cannot open file, path = '/atsign/.ssh/authorized_keys' (OS Error: Permission denied, errno = 13)
```

Broken:

```
chown -R ${USER}:${USER} ${HOMEDIR} ; \
touch ${HOMEDIR}/.ssh/authorized_keys ; \
chmod 600 ${HOMEDIR}/.ssh/authorized_keys
```

Working:

```
touch ${HOMEDIR}/.ssh/authorized_keys ; \
# chmod 600 ${HOMEDIR}/.ssh/authorized_keys ; \
chown -R ${USER}:${USER} ${HOMEDIR}
```

JeremyTubongbanua committed:
feat: working dockerfile

JeremyTubongbanua committed:
feat: chmod `.startup.sh` copy

JeremyTubongbanua commented:
@cpswan 

The current Dockerfile is building and the demo is working well.

Still running into that issue as soon as I go from `FROM ubuntu` to `FROM debian-...`.

Is it okay to merge?

cpswan committed:
fix: Use pinned Ubuntu image

cpswan committed:
build(deps): Configure dependabot to monitor Dockerfile

cpswan commented on sshnp_docker_demo/README.md:
````diff
@@ -0,0 +1,262 @@
+Docker version 23.0.5, build bc4487a
+```
+
+2. You will need 3 atSigns and their associated `.atKeys` files (one for each docker container). For each of your atSigns, put the `.atKeys` file into the `keys/` directory. For example, your file structure should be similar to:`sshnp/keys/@sshnp_key.atKeys`, `sshnpd/keys/@sshnpd_key.atKeys`, `sshrvd/keys/@sshrvd_key.atKeys`.
````
It would be helpful to link to instructions on how to get an atSign, and how to get the .atKeys file for it.

JeremyTubongbanua replied:
Done in https://github.com/atsign-foundation/at_demos/pull/97/commits/9ba4b06c6387eabfa82d03a47f353b2d99984433

Thanks!

cpswan replied:
That video talks a LOT about PicoW. A more generic (text) guide would be better. If we don't have one then let's open a ticket to get one and have a TODO here.

JeremyTubongbanua replied:
I'll find some time to make a new one today

JeremyTubongbanua replied:
written instructions here - https://github.com/atsign-foundation/at_demos/pull/97/commits/752f9689f2175b9c26d49c801db0f6e520cc623c

added a todo for the video coming out soon

cpswan commented on sshnp_docker_demo/db-template.sh:
```diff
@@ -0,0 +1,21 @@
+    exit 1
+fi
+
+rm Dockerfile
```
Test for the file first before trying to delete it to avoid an error message

JeremyTubongbanua replied:
instead, i thought it was a good idea to copy overwrite/force - https://github.com/atsign-foundation/at_demos/pull/97/commits/57a70a1e7227d0dae77eb2703181faa1b6b60977

for the second `rm Dockerfile`, i do the if check - https://github.com/atsign-foundation/at_demos/pull/97/commits/a12d9836483c3e2fa38e65a2fdec37c1e70bb738

cpswan commented on sshnp_docker_demo/db-template.sh:
```diff
@@ -0,0 +1,21 @@
+    echo "Usage: ./db.sh <sshnp/sshnpd/sshrvd>"
+    exit 1
+fi
+
```
Test for presence of .atKeys, otherwise stuff is going to fail later.

JeremyTubongbanua replied:
Done in https://github.com/atsign-foundation/at_demos/pull/97/commits/d516cd0281c05c34ec133573f91aee656f611ec7

cpswan commented on sshnp_docker_demo/db-template.sh:
```diff
@@ -0,0 +1,21 @@
+
+rm Dockerfile
+cp ../Dockerfile .
+docker container prune -f
```
This is pretty rude. People might have images and containers lying around that would be destroyed here.

JeremyTubongbanua replied:
Removed prunes https://github.com/atsign-foundation/at_demos/pull/97/commits/769faa5f523903b90824d6e07539b11de2b2fd62

JeremyTubongbanua replied:
Removed prunes https://github.com/atsign-foundation/at_demos/pull/97/commits/769faa5f523903b90824d6e07539b11de2b2fd62

cpswan commented on sshnp_docker_demo/db-template.sh:
```diff
@@ -0,0 +1,21 @@
+
+rm Dockerfile
+cp ../Dockerfile .
+docker container prune -f
```
All of these `docker` commands assume that it can be run without `sudo`, which isn't a safe assumption

JeremyTubongbanua replied:
Would I just `sudo docker` ?

https://github.com/atsign-foundation/at_demos/pull/97/commits/1aaad8c04a64f11085e9a2619e7c1883c5789d4d

cpswan commented on sshnp_docker_demo/README.md:
````diff
@@ -0,0 +1,262 @@
+
+```sh
+cd sshrvd
+sh db.sh
````
I get that it's called db for Docker Build, but usually that would be DataBase. Maybe use a more explicit name.

Also why `sh db.sh` rather than `./db.sh`?

JeremyTubongbanua replied:
Now uses `./` to execute scripts instead of `sh`
https://github.com/atsign-foundation/at_demos/pull/97/commits/1aaad8c04a64f11085e9a2619e7c1883c5789d4d

cpswan replied:
All the scripts need to be `chmod +x` so that they're runnable

JeremyTubongbanua replied:
@cpswan should the script also chmod +x then?

cpswan replied:
You should be doing `chmod +x` *to* the scripts (so they're executable when cloned from git) not *from* the scripts.

JeremyTubongbanua replied:
done https://github.com/atsign-foundation/at_demos/pull/97/commits/a7d0b6f82621935aa5d41f6fda37e8e907e92b73

and here too (i missed a couple) https://github.com/atsign-foundation/at_demos/pull/97/commits/ae38e9cf78d250d7ae6204d7a36e41c356228bd2

cpswan commented on sshnp_docker_demo/README.md:
```diff
@@ -0,0 +1,262 @@
+
+(A) The `sshnp/.startup.sh` script:
+
+- replace `john@example.com` with your email
```
Can we use an interactive script to do TOKEN substitution?

JeremyTubongbanua replied:
This is what I came up with:

![image](https://github.com/atsign-foundation/at_demos/assets/79019866/dc81ac33-4eac-4e6e-8b39-9f0b44490cef)
![image](https://github.com/atsign-foundation/at_demos/assets/79019866/84b522aa-83ed-444a-817c-e8811ab31c6f)

Generates this:

![image](https://github.com/atsign-foundation/at_demos/assets/79019866/fbbaff9f-21ff-4de8-bcb4-e00edaf2966c)

JeremyTubongbanua replied:
I had some trouble naming the script, though - 
![image](https://github.com/atsign-foundation/at_demos/assets/79019866/d9106056-f31b-4794-8a81-6cfd76ac924a)

cpswan reviewed (changes requested):
I got deeper into testing, which raised a bunch more questions.

Also consider merging [cpswan-dockerfile-tidy](https://github.com/atsign-foundation/at_demos/tree/cpswan-dockerfile-tidy) to get a tidier Dockerfile (which now has a pinned ubuntu image at the start)

JeremyTubongbanua committed:
Merge remote-tracking branch 'upstream/cpswan-dockerfile-tidy' into sshnp_docker_demo

JeremyTubongbanua committed:
chore: remove prunes

JeremyTubongbanua committed:
feat: docker build checks for atKeys files

JeremyTubongbanua committed:
feat: rename db.sh to docker-build.sh and sudo docker

JeremyTubongbanua committed:
feat: Dockerfile buildimage

JeremyTubongbanua committed:
feat: Dockerfile base image

JeremyTubongbanua committed:
feat: `cp -rf` Dockerfile

JeremyTubongbanua committed:
feat: check for dockerfile before `rm`

JeremyTubongbanua committed:
docs: atKeys instructions

JeremyTubongbanua committed:
chore: use <> for token substitution

JeremyTubongbanua committed:
feat: good start for interactive setup for startup scripts

JeremyTubongbanua committed:
feat: dockerfile mkdir ~/sshnp

JeremyTubongbanua committed:
chore: dockerfile comment out ENTRYPOINT

JeremyTubongbanua committed:
Merge remote-tracking branch 'upstream/trunk' into sshnp_docker_demo

JeremyTubongbanua commented:
I commented out

```Dockerfile
ENTRYPOINT [ "sh", "/atsign/.startup.sh" ]
```

because `sshnp` container stops once the ssh command is generated

JeremyTubongbanua committed:
feat: script-templates dir

JeremyTubongbanua committed:
feat: interactive script uses templates

JeremyTubongbanua committed:
chore: gitignore tmp/generated files

JeremyTubongbanua committed:
fix: change docker-build-template location in docker-build scripts

JeremyTubongbanua committed:
fix: sshnpdrvd -> sshrvd

JeremyTubongbanua committed:
fix: remove <> for deviceName interactive script

JeremyTubongbanua committed:
docs: use `./` instead of `sh`

JeremyTubongbanua committed:
docs: use `./` instead of `sh`, refactor `db.sh` to `docker-build.sh`, now uses interactive setup

JeremyTubongbanua commented:
Updated documentation to use
- refactored `docker-build.sh` script
- new interactive shell script

https://github.com/atsign-foundation/at_demos/pull/97/commits/a7d0b6f82621935aa5d41f6fda37e8e907e92b73

JeremyTubongbanua committed:
feat: demo-base/sshnp_docker_demo_base Dockerfile

JeremyTubongbanua committed:
feat: comment entry points

JeremyTubongbanua committed:
docs: sshrvd optional *

JeremyTubongbanua committed:
docs: new image name `sshnp_docker_demo_sshrvd`

JeremyTubongbanua commented:
@cpswan I created a `atsigncompany/sshnp_docker_demo_base` base image. https://github.com/atsign-foundation/at_demos/pull/97/commits/da99ac7aa4e27491ec4d1dfe5e775350366dc85b

The structure I went for is:

- base image: `atsigncompany/sshnp_docker_demo_base`
- sshnp image `atsigncompany/sshnp_docker_demo_sshnp` -> `sshnp_docker_demo_sshnp` container
- sshnpd image `atsigncompany/sshnp_docker_demo_sshnpd` -> `sshnp_docker_demo_sshnpd` container
- sshrvd image `atsigncompany/sshnp_docker_demo_sshrvd` -> `sshrvd_docker_demo_sshrvd` container

JeremyTubongbanua committed:
docs: getting atKeys instructions

JeremyTubongbanua committed:
chore added `sshnp_docker_demo/*/.startup.sh` to `.gitignore`

cpswan commented on sshnp_docker_demo/demo-base/Dockerfile:
```diff
@@ -0,0 +1,48 @@
+    apt-get update ; \
+    apt-get install -y openssh-server wget sudo iputils-ping iproute2 ncat \
+    telnet net-tools nmap iperf3 tmux traceroute nano vim watch curl dpkg ; \
+    case "$(dpkg --print-architecture)" in \
```
Case statement needs to be properly indented for readability

JeremyTubongbanua replied:
https://github.com/atsign-foundation/at_demos/pull/97/commits/579e8989c424e2ef110583e6f46ae63c8cf4be06

cpswan commented on sshnp_docker_demo/demo-base/Dockerfile:
```diff
@@ -0,0 +1,48 @@
+    tar -xvf sshnp-linux-${ARCH}.tgz ; \
+    rm sshnp-linux-${ARCH}.tgz ;
+
+FROM ubuntu:22.04@sha256:dfd64a3b4296d8c9b62aa3309984f8620b98d87e47492599ee20739e8eb54fbf
```
This shouldn't need to be a two stage build. If we're using something as heavy as Ubuntu, then it should be possible to just do everything in place.

JeremyTubongbanua replied:
https://github.com/atsign-foundation/at_demos/pull/97/commits/ba2205149939f2658e5785af5da51c42e02d8599

cpswan reviewed (changes requested):
There should be a GitHub Action workflow to regularly rebuild the base image.

See https://github.com/atsign-foundation/at_server/blob/trunk/.github/workflows/ve_base.yaml for an example

cpswan reviewed (changes requested):
There should be a GitHub Action workflow to regularly rebuild the base image.

See https://github.com/atsign-foundation/at_server/blob/trunk/.github/workflows/ve_base.yaml for an example

JeremyTubongbanua committed:
refactor: dockerfile base case indentation

JeremyTubongbanua committed:
feat: docker one stage build

JeremyTubongbanua committed:
feat: now uses installer scripts, removed `sshrvd` step

JeremyTubongbanua committed:
docs: improvements

JeremyTubongbanua committed:
docs: quick start

JeremyTubongbanua committed:
docs: quick start improvements

JeremyTubongbanua committed:
docs: README examples

JeremyTubongbanua committed:
cd: GHA `sshnp_docker_demo_base.yml`

JeremyTubongbanua committed:
feat: building checks for dockerhub base image

JeremyTubongbanua committed:
docs: atSign activation readme

JeremyTubongbanua committed:
docs: README

JeremyTubongbanua committed:
docs: README

JeremyTubongbanua committed:
docs: READMe uses installer curls

JeremyTubongbanua committed:
feat: no longer installs binaries

JeremyTubongbanua committed:
feat: docker build now creates a separate network

JeremyTubongbanua committed:
feat: uncomment docker pull

JeremyTubongbanua commented:
@cpswan I have 

1. refactored README to use @XavierChanth's work with the sshnpd/sshnp installers
2. sshnp_docker_demo_base GHA
3. docker-build script tries to pull base image from docker hub

```sh
# try to docker pull, if it doesn't work, build local image
sudo docker pull $BASE_IMAGE_NAME
if [ $? -ne 0 ]
then
    echo "Could not pull $BASE_IMAGE_NAME. Building local image."
    sudo docker build -t $BASE_IMAGE_NAME -f ../demo-base/Dockerfile ../demo-base
fi
```
4. Dockerfile no longer installs binaries (as installers do that for you)

cpswan commented on .github/workflows/sshnp_docker_demo_base.yml:
```diff
@@ -0,0 +1,39 @@
+  schedule:
+    # - cron: '0 0 * * *' # every day, 00:00
+    # - cron: '0 0 * * 0' # every week, 00:00 Sunday
+    - cron: '5 * * * *' # every 5 minutes, temporarily set to 5 minutes for testing
```
workflow_dispatch is what we use for testing. This needs to be a sensible weekly parameter before we can merge, or we'll thrash Actions and Docker Hub

JeremyTubongbanua replied:
https://github.com/atsign-foundation/at_demos/pull/97/commits/2b99b71288b803d971dd300bab543cbe91000f7c

cpswan commented on .github/workflows/sshnp_docker_demo_base.yml:
```diff
@@ -0,0 +1,39 @@
+name: VE_base
```
No it's not

JeremyTubongbanua replied:
https://github.com/atsign-foundation/at_demos/pull/97/commits/2b99b71288b803d971dd300bab543cbe91000f7c

cpswan commented on sshnp_docker_demo/demo-base/Dockerfile:
```diff
@@ -0,0 +1,46 @@
+
+ENTRYPOINT sudo service ssh start && bash
+
+# Useful code for manually downloading sshnp binaries, might be useful later
```
Delete it - it's already in the commit history (and elsewhere in our stuff).

JeremyTubongbanua replied:
https://github.com/atsign-foundation/at_demos/pull/97/commits/4b721f322a2d2808650dfb87af9010a0596c3768

cpswan reviewed (changes requested):
Just a few nits, and then I think we're ready to merge

JeremyTubongbanua committed:
docs: `/bin/bash`

JeremyTubongbanua committed:
chore: remove comments in demo base

JeremyTubongbanua committed:
cd: sshnp_docker_demo_base.yml

JeremyTubongbanua committed:
cd: 1130 on friday
